singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(PowerMG)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    return 0;
+  }
+  static inline unsigned long max_scratch_size(unsigned int nelements) { return 0; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"PowerMG Params: "};
+    printf("%s rho0:%e T0:%e G0:%e Cv0:%e E0:%e S0:%e Pmin:%e\n"
+           "non-zero elements in K0toK40 array:\n",
+           st, _rho0, _T0, _G0, _Cv0, _E0, _S0, _Pmin);
+    for (int i = 0; i < 41; i++) {
+      if (_K0toK40[i] * _K0toK40[i] > 0.0) printf("K%i:%e\t", i, _K0toK40[i]);
+    }
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("PowerMG"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  Real _rho0, _T0, _G0, _Cv0, _E0, _S0, _Pmin;
+  static constexpr const int PressureCoeffsK0toK40Size = 41;
+  int _M;
+  Real _K0toK40[PressureCoeffsK0toK40Size];
+  void _CheckPowerMG();
+  void _InitializePowerMG(const Real *expcoeffs);
+  PORTABLE_INLINE_FUNCTION Real _HugPressureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _HugTemperatureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _SN2Mp2(const Real x) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityTemperature(const Real rho, const Real temp) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityInternalEnergy(const Real rho, const Real sie) const;
+};
+
+inline void PowerMG::_CheckPowerMG() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter T0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter Cv0 < 0");
+  }
+  if (_K0toK40[0] < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter K0 < 0");
+  }
+  if (_Pmin >= 0.0) {
+    _Pmin = -1000 * _K0toK40[0];
+#ifndef NDEBUG
+    PORTABLE_WARN(
+        "PowerMG model parameter Pmin not set or positive. Reset to default (-1000*K0)");
+#endif // NDEBUG
+  }
```

Comment:
This strikes me as a little odd... does the fact that this model supports negative pressures imply that it supports tension? I'm not familiar with PowerMG but usually Mie-Gruneisen does not support tension, right?

Reply:
Tension is tricky and it seems more or less everyone has their own way of dealing with it. This is exactly how it is done in ALEGRA and in the Robinson write-up. See the documentation.

Reply:
You have a few other things I do not like in that region, like negative bulk modulus (isothermal).

Reply:
:+1: It's all fine as long as the behavior is documented and is clear.